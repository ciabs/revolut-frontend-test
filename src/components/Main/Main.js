import React, {Component} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';

import {setFromCurrency, setFromValue, setToCurrency, setToValue, exchange, showSuccessModal} from '../../actions';
import {Container} from './Main.styles';
import {convertInputValueToTwoDigitsNumber} from '../../helpers/string.helper';
import SuccessModal from '../SuccessModal/SuccessModal';
import PropTypes from 'prop-types';
import FromComponent from '../ConvertFrom/ConvertFrom';
import ConvertTo from '../ConvertTo/ConvertTo';

class Main extends Component {
  /*
  componentDidMount() {
    this.fromValueInput.focus();
  }
  */

  handleFromValueChange = event => this.props.setFromValueFunc(
    convertInputValueToTwoDigitsNumber(event.target.value)
  );

  handleToValueChange = event => this.props.setToValueFunc(
    convertInputValueToTwoDigitsNumber(event.target.value)
  );

  handleFromCurrencyChange = (event, index, value) => this.props.setFromCurrencyFunc(value);

  handleToCurrencyChange = (event, index, value) => this.props.setToCurrencyFunc(value);

  handleExchange = () => this.props.exchangeFunc();

  handleHideModal = () => this.props.showSuccessModalFunc(false);

  fixFocus = () => {
    if (this.state.inputOnFocus === 'fromCurrency') {
      this.fromValueInput.focus();
    } else {
      this.toValueInput.focus();
    }
  };

  render() {
    const {rates, active, balance, modal} = this.props;
    const {fromValue, fromCurrency, toValue, toCurrency} = active;
    const isExchangeButtonDisabled = balance[fromCurrency] < fromValue ||
      balance[fromCurrency] === 0 ||
      fromValue === 0;

    const renderSuccessModal = modal.success && (
      <SuccessModal
        handleHideModal={this.handleHideModal}
        lastExchange={balance.lastExchange}
      />
    );

    return (
      <Container>
        {renderSuccessModal}
        <FromComponent
          balance={balance}
          fromCurrency={fromCurrency}
          fromValue={fromValue}
          toCurrency={toCurrency}
          rates={rates}
          handleFromCurrencyChange={this.handleFromCurrencyChange}
          handleFromValueChange={this.handleFromValueChange}
        />
        <ConvertTo
         toCurrency={toCurrency}
         balance={balance}
         toValue={toValue}
         handleToCurrencyChange={this.handleToCurrencyChange}
         handleToValueChange={this.handleToValueChange}
         isExchangeButtonDisabled={isExchangeButtonDisabled}
         rates={rates}
         handleExchange={this.handleExchange}
        />
      </Container>
    );
  }
}

Main.propTypes = {
  rates: PropTypes.object.isRequired,
  active: PropTypes.object.isRequired,
  balance: PropTypes.object.isRequired,
  modal: PropTypes.object.isRequired,
  setFromCurrencyFunc: PropTypes.func.isRequired,
  setFromValueFunc: PropTypes.func.isRequired,
  setToCurrencyFunc: PropTypes.func.isRequired,
  setToValueFunc: PropTypes.func.isRequired,
  exchangeFunc: PropTypes.func.isRequired,
  showSuccessModalFunc: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  rates: state.rates,
  active: state.active,
  balance: state.balance,
  modal: state.modal
});

const mapDispatchToProps = dispatch => {
  return {
    setFromCurrencyFunc: bindActionCreators(setFromCurrency, dispatch),
    setFromValueFunc: bindActionCreators(setFromValue, dispatch),
    setToCurrencyFunc: bindActionCreators(setToCurrency, dispatch),
    setToValueFunc: bindActionCreators(setToValue, dispatch),
    exchangeFunc: bindActionCreators(exchange, dispatch),
    showSuccessModalFunc: bindActionCreators(showSuccessModal, dispatch),
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Main);
