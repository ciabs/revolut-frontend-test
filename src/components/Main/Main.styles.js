import styled from 'styled-components';

import {media} from '../../styles/mediaQuery';

export const Container = styled.div`
  width: 375px;
  height: 667px;
  position: relative;
  ${media.small`
    height: 100vh;
  `};
`;

export const FormWrapper = styled.div`
  display: flex;
  justify-content: space-between;
`;
