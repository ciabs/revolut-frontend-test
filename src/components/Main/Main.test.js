import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';

import {Main} from './Main';

describe('Main', () => {
  const props = {
    rates: {},
    active: {
      fromValue: 1,
      fromCurrency: 'EUR',
      toValue: 2,
      toCurrency: 'GBP'
    },
    balance: {
      GBP: 15.00,
      EUR: 0,
      USD: 0
    },
    modal: {},
    setFromCurrencyFunc: jest.fn(),
    setFromValueFunc: jest.fn(),
    setToCurrencyFunc: jest.fn(),
    setToValueFunc: jest.fn(),
    exchangeFunc: jest.fn(),
    showSuccessModalFunc: jest.fn(),
    errors: {
      rates: null
    },
    loading: {
      rates: false
    }
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MuiThemeProvider>
        <Main {...props} />
      </MuiThemeProvider>,
      div
    );
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <MuiThemeProvider>
        <Main {...props} />
      </MuiThemeProvider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
