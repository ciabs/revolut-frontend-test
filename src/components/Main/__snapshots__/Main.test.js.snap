// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main has a valid snapshot 1`] = `
<div
  className="sc-bdVaJa ehbYjZ"
>
  <div
    className="sc-iwsKbI gpzLhp"
  >
    <span
      aria-hidden={true}
      className="fa fa-exclamation-triangle fa-2x"
      style={
        Object {
          "color": "red",
          "cursor": "pointer",
          "display": "none",
          "position": "absolute",
          "right": "20px",
          "top": "20px",
        }
      }
      title={null}
    />
    <div
      className="sc-bwzfXH MKmUa"
    >
      <div
        className={undefined}
        name="fromCurrency"
        style={
          Object {
            "backgroundColor": "transparent",
            "cursor": "not-allowed",
            "display": "inline-block",
            "fontFamily": "Roboto, sans-serif",
            "fontSize": 16,
            "height": 48,
            "lineHeight": "24px",
            "muiPrepared": true,
            "position": "relative",
            "transition": "height 200ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "width": 256,
          }
        }
      >
        <div
          style={
            Object {
              "bottom": 12,
              "color": "rgba(0, 0, 0, 0.3)",
              "muiPrepared": true,
              "opacity": 0,
              "position": "absolute",
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            }
          }
        >
           
        </div>
        <div
          className={undefined}
          id="fromCurrency"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          style={
            Object {
              "WebkitOpacity": 1,
              "WebkitTapHighlightColor": "rgba(0,0,0,0)",
              "backgroundColor": "rgba(0,0,0,0)",
              "border": "none",
              "color": "rgba(0, 0, 0, 0.3)",
              "cursor": "inherit",
              "display": "block",
              "font": "inherit",
              "fontFamily": "Roboto, sans-serif",
              "fontSize": 15,
              "height": "100%",
              "muiPrepared": true,
              "outline": "none",
              "padding": 0,
              "position": "relative",
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "width": "100%",
            }
          }
        >
          <div
            onClick={[Function]}
            style={
              Object {
                "cursor": "not-allowed",
                "height": "100%",
                "muiPrepared": true,
                "position": "relative",
                "width": "100%",
              }
            }
          >
            <div
              style={
                Object {
                  "boxSizing": "border-box",
                  "content": "' '",
                  "display": "table",
                  "muiPrepared": true,
                }
              }
            />
            <div
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.3)",
                  "fontSize": "30px",
                  "height": "56px",
                  "lineHeight": "56px",
                  "muiPrepared": true,
                  "opacity": 1,
                  "overflow": "hidden",
                  "paddingLeft": 0,
                  "paddingRight": 56,
                  "position": "relative",
                  "textOverflow": "ellipsis",
                  "top": -4,
                  "whiteSpace": "nowrap",
                }
              }
            >
              EUR
            </div>
            <button
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseOut={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
                  "background": "none",
                  "border": 10,
                  "boxSizing": "border-box",
                  "cursor": "default",
                  "display": "inline-block",
                  "fill": "#e0e0e0",
                  "fontFamily": "Roboto, sans-serif",
                  "fontSize": 0,
                  "fontWeight": "inherit",
                  "height": 48,
                  "margin": 0,
                  "muiPrepared": true,
                  "outline": "none",
                  "overflow": "visible",
                  "padding": 12,
                  "position": "absolute",
                  "right": 0,
                  "textDecoration": "none",
                  "top": 0,
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                  "verticalAlign": null,
                  "width": 48,
                }
              }
              tabIndex={-1}
              type="button"
            >
              <svg
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={
                  Object {
                    "color": "rgba(0, 0, 0, 0.3)",
                    "cursor": "default",
                    "display": "inline-block",
                    "fill": "rgba(0, 0, 0, 0.3)",
                    "height": 24,
                    "muiPrepared": true,
                    "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                    "userSelect": "none",
                    "width": 24,
                  }
                }
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </button>
            <div
              style={
                Object {
                  "borderTop": "none",
                  "bottom": 1,
                  "left": 0,
                  "margin": "-1px 24px",
                  "muiPrepared": true,
                  "position": "absolute",
                  "right": 0,
                }
              }
            />
            <div
              style={
                Object {
                  "boxSizing": "border-box",
                  "clear": "both",
                  "content": "' '",
                  "display": "table",
                  "muiPrepared": true,
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "display": "none",
              }
            }
          />
        </div>
        <div>
          <hr
            aria-hidden="true"
            style={
              Object {
                "borderBottomStyle": "dotted",
                "borderBottomWidth": 2,
                "borderColor": "rgba(0, 0, 0, 0.3)",
                "borderLeft": "none",
                "borderRight": "none",
                "borderTop": "none",
                "bottom": 8,
                "boxSizing": "content-box",
                "margin": 0,
                "muiPrepared": true,
                "position": "absolute",
                "width": "100%",
              }
            }
          />
          <hr
            aria-hidden="true"
            style={
              Object {
                "borderBottomStyle": "solid",
                "borderBottomWidth": 2,
                "borderColor": "#00bcd4",
                "borderLeft": "none",
                "borderRight": "none",
                "borderTop": "none",
                "bottom": 8,
                "boxSizing": "content-box",
                "margin": 0,
                "muiPrepared": true,
                "position": "absolute",
                "transform": "scaleX(0)",
                "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                "width": "100%",
              }
            }
          />
        </div>
      </div>
      <input
        className="sc-bZQynM cJWkqu"
        disabled={true}
        min="0"
        name="fromValue"
        onChange={[Function]}
        placeholder="0"
        type="number"
        value={-1}
      />
    </div>
    <div
      className="sc-gzVnrw kDpknI"
    >
      Balance: 
      â‚¬
      0.00
    </div>
    <div
      className="sc-htoDjs lezqkm"
    >
      <div>
        <span
          aria-hidden={true}
          className="fa fa-circle-o-notch fa-spin"
        />
      </div>
    </div>
  </div>
  <div
    className="sc-gZMcBi dPQFrM"
  >
    <div
      className="sc-bwzfXH MKmUa"
    >
      <div
        className={undefined}
        name="toCurrency"
        style={
          Object {
            "backgroundColor": "transparent",
            "cursor": "not-allowed",
            "display": "inline-block",
            "fontFamily": "Roboto, sans-serif",
            "fontSize": 16,
            "height": 48,
            "lineHeight": "24px",
            "muiPrepared": true,
            "position": "relative",
            "transition": "height 200ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "width": 256,
          }
        }
      >
        <div
          style={
            Object {
              "bottom": 12,
              "color": "rgba(0, 0, 0, 0.3)",
              "muiPrepared": true,
              "opacity": 0,
              "position": "absolute",
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            }
          }
        >
           
        </div>
        <div
          className={undefined}
          id="toCurrency"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          style={
            Object {
              "WebkitOpacity": 1,
              "WebkitTapHighlightColor": "rgba(0,0,0,0)",
              "backgroundColor": "rgba(0,0,0,0)",
              "border": "none",
              "color": "rgba(0, 0, 0, 0.3)",
              "cursor": "inherit",
              "display": "block",
              "font": "inherit",
              "fontFamily": "Roboto, sans-serif",
              "fontSize": 15,
              "height": "100%",
              "muiPrepared": true,
              "outline": "none",
              "padding": 0,
              "position": "relative",
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "width": "100%",
            }
          }
        >
          <div
            onClick={[Function]}
            style={
              Object {
                "cursor": "not-allowed",
                "height": "100%",
                "muiPrepared": true,
                "position": "relative",
                "width": "100%",
              }
            }
          >
            <div
              style={
                Object {
                  "boxSizing": "border-box",
                  "content": "' '",
                  "display": "table",
                  "muiPrepared": true,
                }
              }
            />
            <div
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.3)",
                  "fontSize": "30px",
                  "height": "56px",
                  "lineHeight": "56px",
                  "muiPrepared": true,
                  "opacity": 1,
                  "overflow": "hidden",
                  "paddingLeft": 0,
                  "paddingRight": 56,
                  "position": "relative",
                  "textOverflow": "ellipsis",
                  "top": -4,
                  "whiteSpace": "nowrap",
                }
              }
            >
              GBP
            </div>
            <button
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseOut={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
                  "background": "none",
                  "border": 10,
                  "boxSizing": "border-box",
                  "cursor": "default",
                  "display": "inline-block",
                  "fill": "#e0e0e0",
                  "fontFamily": "Roboto, sans-serif",
                  "fontSize": 0,
                  "fontWeight": "inherit",
                  "height": 48,
                  "margin": 0,
                  "muiPrepared": true,
                  "outline": "none",
                  "overflow": "visible",
                  "padding": 12,
                  "position": "absolute",
                  "right": 0,
                  "textDecoration": "none",
                  "top": 0,
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                  "verticalAlign": null,
                  "width": 48,
                }
              }
              tabIndex={-1}
              type="button"
            >
              <svg
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={
                  Object {
                    "color": "rgba(0, 0, 0, 0.3)",
                    "cursor": "default",
                    "display": "inline-block",
                    "fill": "rgba(0, 0, 0, 0.3)",
                    "height": 24,
                    "muiPrepared": true,
                    "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                    "userSelect": "none",
                    "width": 24,
                  }
                }
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </button>
            <div
              style={
                Object {
                  "borderTop": "none",
                  "bottom": 1,
                  "left": 0,
                  "margin": "-1px 24px",
                  "muiPrepared": true,
                  "position": "absolute",
                  "right": 0,
                }
              }
            />
            <div
              style={
                Object {
                  "boxSizing": "border-box",
                  "clear": "both",
                  "content": "' '",
                  "display": "table",
                  "muiPrepared": true,
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "display": "none",
              }
            }
          />
        </div>
        <div>
          <hr
            aria-hidden="true"
            style={
              Object {
                "borderBottomStyle": "dotted",
                "borderBottomWidth": 2,
                "borderColor": "rgba(0, 0, 0, 0.3)",
                "borderLeft": "none",
                "borderRight": "none",
                "borderTop": "none",
                "bottom": 8,
                "boxSizing": "content-box",
                "margin": 0,
                "muiPrepared": true,
                "position": "absolute",
                "width": "100%",
              }
            }
          />
          <hr
            aria-hidden="true"
            style={
              Object {
                "borderBottomStyle": "solid",
                "borderBottomWidth": 2,
                "borderColor": "#00bcd4",
                "borderLeft": "none",
                "borderRight": "none",
                "borderTop": "none",
                "bottom": 8,
                "boxSizing": "content-box",
                "margin": 0,
                "muiPrepared": true,
                "position": "absolute",
                "transform": "scaleX(0)",
                "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                "width": "100%",
              }
            }
          />
        </div>
      </div>
      <input
        className="sc-bZQynM ksAmJg"
        disabled={true}
        min="0"
        name="toValue"
        onChange={[Function]}
        placeholder="0"
        type="number"
        value={2}
      />
    </div>
    <div
      className="sc-gzVnrw kDpknI"
    >
      Balance: 
      Â£
      15.00
    </div>
    <button
      className="sc-EHOje loVsjv"
      disabled={true}
      onClick={[Function]}
    >
      Exchange
    </button>
  </div>
</div>
`;
